scripts/gen_gdt.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2017 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import sys
+import struct
+import os
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import SymbolTableSection
+
+def debug(text):
+    if not args.verbose:
+        return
+    sys.stdout.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+
+def error(text):
+    sys.stderr.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+    sys.exit(1)
+
+
+gdt_pd_fmt = "<HIH"
+
+FLAGS_GRAN = 1 << 7 # page granularity
+ACCESS_EX = 1 << 3  # executable
+ACCESS_DC = 1 << 2  # direction/conforming
+ACCESS_RW = 1 << 1  # read or write permission
+
+# 6 byte pseudo descriptor, but we're going to actually use this as the
+# zero descriptor and return 8 bytes
+def create_gdt_pseudo_desc(addr, size):
+    # ...and take back one byte for the Intel god whose Ark this is...
+    size = size - 1
+    return struct.pack(gdt_pd_fmt, size, addr, 0)
+
+
+# Limit argument always in bytes
+def chop_base_limit(base, limit):
+    base_lo = base & 0xFFFF
+    base_mid = (base >> 16) & 0xFF
+    base_hi = (base >> 24) & 0xFF
+
+    limit_lo = limit & 0xFFFF
+    limit_hi = (limit >> 16) & 0xF
+
+    return (base_lo, base_mid, base_hi, limit_lo, limit_hi)
+
+
+gdt_ent_fmt = "<HHBBBB"
+
+def create_code_data_entry(base, limit, dpl, flags, access):
+    base_lo, base_mid, base_hi, limit_lo, limit_hi = chop_base_limit(base,
+            limit)
+
+    # This is a valid descriptor
+    present = 1
+
+    # 32-bit protected mode
+    size = 1
+
+    # 1 = code or data, 0 = system type
+    desc_type = 1
+
+    # Just set accessed to 1 already so the CPU doesn't need it update it,
+    # prevents freakouts if the GDT is in ROM, we don't care about this
+    # bit in the OS
+    accessed = 1
+
+    access = access | (present << 7) | (desc_type << 4) | accessed
+    flags = flags | (size << 6) | limit_hi
+
+    return struct.pack(gdt_ent_fmt, limit_lo, base_lo, base_mid,
+                       access, flags, base_hi)
+
+
+def create_tss_entry(base, limit, dpl):
+    present = 1
+
+    base_lo, base_mid, base_hi, limit_lo, limit_hi, = chop_base_limit(base,
+            limit)
+
+    type_code = 0x9 # non-busy 32-bit TSS descriptor
+    gran = 0
+
+    flags = (gran << 7) | limit_hi
+    type_byte = ((present << 7) | (dpl << 5) | type_code)
+
+
+    return struct.pack(gdt_ent_fmt, limit_lo, base_lo, base_mid,
+                       type_byte, flags, base_hi)
+
+
+
+def get_symbols(elf):
+    symbols = elf.get_section_by_name(".symtab")
```

Comment:
Will this work with all combinations of Python/pyelftools? Maybe iterating over the section table looking for some SymbolTableSection instance is going to be more robust.  The gen_offsets.py script has a function you can copy.  Since it raises an exception if the symbol table is absent from the ELF image, the next if can be junked as well.